### PR TITLE
[core] Migrate `FactoryRegistry` to API 2.0

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Integrate `FactoryRegistry` into `ov::OpSet`.
 - Remove `FactoryAttributeAdapter`.
 - Move `ov::OpSet` non-template members implementation to cpp.

### Tickets:
 - [CVS-130103](https://jira.devtools.intel.com/browse/CVS-130103)

### CODE REVIEW

File: src/core/include/openvino/opsets/factory.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#include "openvino/core/visibility.hpp"
+
+namespace ov {
+
+/**
+ * @brief Registry of factories that can construct objects derived from BASE_TYPE.
+ */
+template <typename BASE_TYPE>
+class FactoryRegistry {
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/opsets/factory.hpp:18]: as far as I remember, Factory was supposed to be moved to private space, but don't remember details. 
Do you think it's possible ?

File: src/core/include/openvino/opsets/factory.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <functional>
+#include <unordered_map>
+
+#include "openvino/core/visibility.hpp"
```
Comment (vurusovs) [inline src/core/include/openvino/opsets/factory.hpp:10]: is it used?

File: src/core/include/openvino/opsets/factory.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
```
Comment (vurusovs) [inline src/core/include/openvino/opsets/factory.hpp:5]: I suppose we may make `factory.hpp` private for users and move out from `include/`

@ilya-lavrenov

**Review by vurusovs — CHANGES_REQUESTED** [review]

File: src/core/include/openvino/opsets/factory.hpp (hunk 1 above)
Comment (vurusovs) [reply src/core/include/openvino/opsets/factory.hpp:18]: `ov::OpSet` is only the consumer of `FactoryRegistry`, so I guess we may hide in `opset.cpp` or some other place

File: src/core/include/openvino/opsets/factory.hpp (hunk 3 above)
Comment (praasz) [reply src/core/include/openvino/opsets/factory.hpp:5]: Is it used in test code as class also.
And also OpSet class use template function for insert (header only) and use this factory  template member.

File: src/core/include/openvino/opsets/factory.hpp (hunk 2 above)
Comment (praasz) [reply src/core/include/openvino/opsets/factory.hpp:10]: Will remove it.

File: src/core/include/openvino/opsets/factory.hpp (hunk 1 above)
Comment (praasz) [reply src/core/include/openvino/opsets/factory.hpp]: Remove `FactoryRegistry` class and integrate its most important features into `OpSet` as protected/private part.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/include/ngraph/factory_adapter.hpp [hunk 4]
```diff
@@ -17,7 +17,7 @@
 #include "ngraph/attribute_adapter.hpp"
 #include "ngraph/attribute_visitor.hpp"
 #include "ngraph/deprecated.hpp"
-#include "ngraph/factory.hpp"
```
Comment (vurusovs) [inline src/core/include/ngraph/factory_adapter.hpp:20]: Will the file be deleted?

File: src/core/tests/visitors/op/group_normalization.cpp [hunk 5]
```diff
@@ -14,7 +14,7 @@ using namespace ov;
 using ov::test::NodeBuilder;
 
 TEST(attributes, group_normalization) {
-    NodeBuilder::get_ops().register_factory<ov::op::v10::Unique>();
+    NodeBuilder::opset().insert<ov::op::v12::GroupNormalization>();
```
Comment (t-jankowski) [inline src/core/tests/visitors/op/group_normalization.cpp:17]: Didn't it change the test result? The changed itself is fine - I'm just curious.

**Review by t-jankowski — APPROVED** [review]

File: src/core/tests/visitors/op/group_normalization.cpp (hunk 5 above)
Comment (praasz) [reply src/core/tests/visitors/op/group_normalization.cpp:17]: Without change this was test error as builder has not this operator in opset and cand build, visit attributes for tested operator. 

Previous implementation  was using `FactoryRegistry` which on first use load one of opset table and operator could be in factory map. New implementation is using `OpSet` class which "test_opset" which has empty factory map.
This change detect test issues not only for this operator.

**Pull request merged**